add extractors for fantia and fanbox

thatfuckingbird opened the pull request:
Implemented post and user extractors for Fantia and Fanbox. Both use cookies for auth.
Doing it in 1 PR because these are pretty similar.

I added tests for both, however I don't know how to pass cookies to test_results.py so I didn't manage to run the Fantia tests.
The tests use freely available content, but registration (and for Fantia, subscription to the free plan) is likely needed.

Note that the test urls contain NSFW content (it's hard to find freely available + sfw on these sites).

Closes #849.
Closes #1260. 
Closes #739.

thatfuckingbird commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+
+        for url, data in self.posts():
+            yield Message.Directory, data
+            yield Message.Url, url, data
```
not sure this is ok with yielding both dir and url for all posts, but i dont know how else to do it

mikf replied:
It is OK, but it would be better to use one `Message.Directory` for each post and a `Message.Url` for each file in that post.

mikf replied:
If possible, this should be rewritten so that `self.posts()` yields actual post objects (`pbody` from line 53).
Those would be used in one `yield Message.Directory, post` each, and further down you'd extract the actual files and download them with `yield Message.Url, file_url, post`.
Basically `_get_post_data_and_urls()` should be split into two parts.

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+        while url:
+            url = text.ensure_http_scheme(url)
+            body = self.request(url, headers=headers).json()["body"]
+            for item in body["items"]:
+                for url, data in self._get_post_data_and_urls(item["id"]):
+                    yield url, data
```
This should be done with `yield from ...`

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+        url = "https://api.fanbox.cc/post.info?postId="+post_id
+        pbody = self.request(url, headers=headers).json()["body"]
+
+        post = {
+            "id": pbody["id"],
+            "title": pbody["title"],
+            "fee_required": pbody["feeRequired"],
+            "published": pbody["publishedDatetime"],
+            "updated": pbody["updatedDatetime"],
+            "type": pbody["type"],
+            "tags": pbody["tags"],
+            "creator_id": pbody["creatorId"],
+            "has_adult_content": pbody["hasAdultContent"],
+            "post_url": self.root+"/@"+pbody["creatorId"]+"/posts/"+post_id,
+            "creator_user_id": pbody["user"]["userId"],
+            "creator_name": pbody["user"]["name"],
+            "text": pbody["body"].get("text", None) if pbody["body"] else None,
+            "is_cover_image": False
+        }
```
Why don't we use the "native" post object (`pbody`) and modify it slightly, instead of creating a new one while renaming most of its fields?

thatfuckingbird replied:
mostly it was to use snake_case instead of camelCase but if that is not a concern then I agree it's better to leave it, refactored accordingly

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+            "id": pbody["id"],
+            "title": pbody["title"],
+            "fee_required": pbody["feeRequired"],
+            "published": pbody["publishedDatetime"],
```
Parse `pbody["publishedDatetime"]` with `text.parse_datetime()` and add it as a `date` field.

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+        }
+
+        num = 0
+        if "coverImageUrl" in pbody and pbody["coverImageUrl"]:
```
```suggestion
        cover_image = pbody.get("coverImageUrl")
        if cover_image:
```
and replace all `pbody["coverImageUrl"]` with `cover_image`

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+
+        num = 0
+        if "coverImageUrl" in pbody and pbody["coverImageUrl"]:
+            final_post = dict(post)
```
```suggestion
            final_post = post.copy()
```

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+            final_post = dict(post)
+            final_post["isCoverImage"] = True
+            final_post["file_url"] = pbody["coverImageUrl"]
+            final_post = text.nameext_from_url(
+                pbody["coverImageUrl"], final_post
+            )
```
```suggestion
            text.nameext_from_url(cover_image, final_post)
```
`nameext_from_url` modifies its second argument in-place

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+            num += 1
+            yield pbody["coverImageUrl"], final_post
+
+        for group in ["images", "imageMap"]:
```
```suggestion
        for group in ("images", "imageMap"):
```
That's just a personal pet peeve, but I don't like using lists when tuples work as well.

rautamiekka replied:
Tuples take less time to create, and when you ain't modifying the list there just ain't any point in using lists instead of tuples.

So it's more than just a personal pet peeve, it's actually a better way.

mikf replied:
You are correct, but the difference between the two is negligible (<10%) and is therefore not really important. The difference in runtime between `dict(post)` and `post.copy()`, for example, is a lot more significant in comparison (~50%), but even that doesn't matter when compared to anything I/O related.

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+                    )
+                    if "extension" in item:
+                        final_post["extension"] = item["extension"]
+                    final_post["file_id"] = item.get("id", None)
+                    final_post["width"] = item.get("width", None)
+                    final_post["height"] = item.get("height", None)
```
`item.get()` returns `None` by default if the field doesn't exist

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,166 @@
+class FanboxCreatorExtractor(FanboxExtractor):
+    """Extractor for a fanbox creator's works"""
+    subcategory = "creator"
+    pattern = (r"(?:https?://)?([a-zA-Z0-9_-]+)\.fanbox\.cc/?$|"
+               r"(?:https?://)?(?:www\.)?fanbox\.cc/@([^/?#]+)/?$")
```
I think it'd be better to use a `BASE_PATTERN` like [here](https://github.com/mikf/gallery-dl/blob/78d7ee3ef4eed38b24591fcd553bb3a08e3215db/gallery_dl/extractor/deviantart.py#L21-L25) and [here](https://github.com/mikf/gallery-dl/blob/78d7ee3ef4eed38b24591fcd553bb3a08e3215db/gallery_dl/extractor/deviantart.py#L373)

mikf commented on gallery_dl/extractor/fantia.py:
```diff
@@ -0,0 +1,154 @@
+        for url, data in self.posts():
+            if data["content_filename"]:
+                fname = data["content_filename"]
+                data["filename"] = fname
+                if "." in fname:
+                    data["extension"] = fname[fname.rfind(".")+1:]
+                else:
+                    data["extension"] = text.ext_from_url(url)
```
`filename` should not include the filename extension.
Why not `text.nameext_from_url(data["content_filename"] or url, data)`?

mikf commented on gallery_dl/extractor/fantia.py:
```diff
@@ -0,0 +1,154 @@
+                gallery_page_html, 'class="link-block" href="/posts/', '"'
+            ):
+                posts_found = True
+                for url, data in self._get_post_data_and_urls(post_id):
+                    yield url, data
```
`yield from` like before, but it'd be better to split `_get_post_data_and_urls()`

mikf commented on gallery_dl/extractor/fantia.py:
```diff
@@ -0,0 +1,154 @@
+class FantiaCreatorExtractor(FantiaExtractor):
+    """Extractor for a fantia creator's works"""
+    subcategory = "creator"
+    pattern = r"(?:https?://)?(?:www\.)?fantia\.jp/fanclubs/([^/?#]+)"
```
Use `\d` instead of `[^/?#]` if a fanclub ID is always just a number,

mikf commented on gallery_dl/extractor/fantia.py:
```diff
@@ -0,0 +1,154 @@
+class FantiaPostExtractor(FantiaExtractor):
+    """Extractor for media from a single fantia post"""
+    subcategory = "post"
+    pattern = r"(?:https?://)?(?:www\.)?fantia\.jp/posts/([^/?#]+)"
```
same

mikf reviewed (commented):
A huge Thank You for adding extractors for both sites to you!
I hope I'm not too nitpicky about some this ...

thatfuckingbird commented:
> I hope I'm not too nitpicky about some this ...

Don't worry about it, it's helpful.

thatfuckingbird commented:
ok, I think everything above is addressed

abslamp commented:
You you plan to support saving video links (maybe as text) in Fanbox? I think for some posts the `type` in the json is "video", and it contains a "video" key in body's body, with content like: `"video":{"serviceProvider":"youtube","videoId":"youtubeId"}`. Maybe useful to invoke a post action like youtube-dl.

thatfuckingbird commented:
@abslamp Sure, I didn't know fanbox had this feature. If you can share the whole post response (with the paid urls removed) then I can add it.

abslamp commented:
> 
> 
> @abslamp Sure, I didn't know fanbox had this feature. If you can share the whole post response (with the paid urls removed) then I can add it.

@thatfuckingbird Thanks for considering this!

URL: https://www.fanbox.cc/@ayumasayu/posts/1737774

Formatted JSON with paid contents removed:
```
{
   "body":{
      "id":"1737774",
      "title":"\u7483\u5948\u3061\u3083\u3093\u30bf\u30a4\u30e0\u30e9\u30d7\u30b9",
      "coverImageUrl":null,
      "feeRequired":864,
      "publishedDatetime":"2020-12-28T17:51:48+09:00",
      "updatedDatetime":"2021-01-02T09:31:12+09:00",
      "type":"video",
      "body":{
         "text":[REDUCTED],
         "video":{
            "serviceProvider":"youtube",
            "videoId":[REDUCTED, the last part of youtube's video url]
         }
      },
      "tags":[
         
      ],
      "excerpt":[REDUCTED],
      "isLiked":false,
      "likeCount":18,
      "commentCount":1,
      "restrictedFor":null,
      "user":{
         "userId":"2473967",
         "name":"\u3042\u3086\u307e\u7d17\u7531",
         "iconUrl":"https:\/\/pixiv.pximg.net\/c\/160x160_90_a2_g5\/fanbox\/public\/images\/user\/2473967\/icon\/fA0s7GFQgdcE21pwISHHmzvg.jpeg"
      },
      "creatorId":"ayumasayu",
      "hasAdultContent":true,
      "commentList": [REDUCTED]
      "nextPost":{
         "id":"1745343",
         "title":"\u306e\u3063\u304b\u308a\u3042\u3059\u3061\u3083\u3093",
         "publishedDatetime":"2020-12-30 19:24:10"
      },
      "prevPost":{
         "id":"1734302",
         "title":"\u8868\u7d19\u69cb\u56f3\u6848\u30e9\u30d5",
         "publishedDatetime":"2020-12-27 17:53:37"
      },
      "imageForShare":"https:\/\/pixiv.pximg.net\/c\/1200x630_90_a2_g5\/fanbox\/public\/images\/creator\/2473967\/cover\/02RjvKWibZct6SWzEqUJ5Sms.jpeg"
   }
}
```
Please note it does not provide the whole URL, but breaking it into 2 parts. For example, for `https://www.youtube.com/watch?v=VjHMmF7HsoU` (a random public video appeared in my youtube's suggestions), it would be like:
```
"video":{
      "serviceProvider":"youtube",
      "videoId":"VjHMmF7HsoU"
}
```

thatfuckingbird commented:
@abslamp Hmm yea I would just add the "video" object to the metadata, if present. Problem is, in this example gallery-dl would produce no files for this post, so there would be no metadata written at all in the end...

abslamp commented:
> 
> 
> @abslamp Hmm yea I would just add the "video" object to the metadata, if present. Problem is, in this example gallery-dl would produce no files for this post, so there would be no metadata written at all in the end...
> 
> Maybe I could yield the youtube URL and worst case it could be captured with the 'write unsupported URLs to file' option, but that won't work for other serviceProviders....

@thatfuckingbird You can use this page for testing. But I don't suggest you adding this to test because I will remove this page later. https://mkx-bot.fanbox.cc/posts/2130801

From the creator panel I found that currently fanbox supports YouTube, Vimeo, and SoundCloud only.

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,157 @@
+    def posts(self):
+        """Return all relevant post objects"""
+
+    def _pagination(self, url):
+        headers = {"Origin": self.root}
+
+        while url:
+            url = text.ensure_http_scheme(url)
+            body = self.request(url, headers=headers).json()["body"]
+            for item in body["items"]:
+                yield self._get_post_data(item["id"])
+
+            url = body["nextUrl"]
```
The items returned from `https://api.fanbox.cc/post.listCreator?creatorId=USER&limit=10` appear to be, at least for `xub.fanbox.cc`, more or less the same as the single-item results from `https://api.fanbox.cc/post.info?postId=ID` (*) It's only missing comments, `imageForShare`, and the entries about next/previous posts.

If that's true in general (posts with videos, embeds, etc), we don't need to fetch data from `/post.info` for every post and can use the items returned from `/post.listCreator` directly.

<details>
 <summary>(*) Diff "/post.listCreator" - "/post.info" for post 2059366</summary>

``` diff
   "creatorId": "xub",
-  "hasAdultContent": true,
-  "commentList": {
-    "items": [],
-    "nextUrl": null
-  },
-  "nextPost": {
-    "id": "2085876",
-    "title": "Skeb Commission",
-    "publishedDatetime": "2021-04-03 04:56:12"
-  },
-  "prevPost": {
-    "id": "2009099",
-    "title": "ﾒｽｶﾞｷ〇〇★〇〇〇〇〇をわからせる絵",
-    "publishedDatetime": "2021-03-12 19:01:08"
-  },
-  "imageForShare": "https://pixiv.pximg.net/c/1200x630_90_a2_g5/fanbox/public/images/post/2059366/cover/JuTXNtvo1BRN93cLW371vVd6.jpeg"
+  "hasAdultContent": true
 }

```
</details>

thatfuckingbird replied:
Seems like you are right, at least I couldn't find any posts where listCreator didn't have all the content. Updated the code to use that directly, it's easy enough to change it back if it turns out it's needed.

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,157 @@
+class FanboxCreatorExtractor(FanboxExtractor):
+    """Extractor for a Fanbox creator's works"""
+    subcategory = "creator"
+    pattern = BASE_PATTERN + r"/?$"
```
I think this should also match `/posts` URLs, e.g. `https://xub.fanbox.cc/posts`
```suggestion
    pattern = BASE_PATTERN + r"(?:/posts)?/?$"
```

mikf commented on gallery_dl/extractor/fantia.py:
```diff
@@ -0,0 +1,149 @@
+    def posts(self):
+        """Return all relevant post objects"""
+
+    def _pagination(self, base_url):
+        headers = {"Referer": self.root}
+        page = 1
+        posts_found = True
+
+        while posts_found:
+            url = base_url+str(page)
+            url = text.ensure_http_scheme(url)
+            gallery_page_html = self.request(url, headers=headers).text
+            posts_found = False
+            for post_id in text.extract_iter(
+                gallery_page_html, 'class="link-block" href="/posts/', '"'
+            ):
+                posts_found = True
+                yield self._get_post_data(post_id)
+
+            page += 1
```
I'd suggest using separate `url` and `params` instead of using a `base_url` and manually adding the current page. This could theoretically also be done for Fanbox, but we have `body["nextUrl"]` there.
I'd also only return post IDs from here (and all `posts()` methods including the one from `FantiaPostExtractor`) and call  `_get_post_data(post_id)` in the main loop. We can also get rid of `posts_found` and the call to `ensure_http_scheme()`.

```suggestion
def _pagination(self, url):
        params = {"page": 1}
        headers = {"Referer": self.root}

        while True:
            page = self.request(url, params=params, headers=headers).text

            post_id = None
            for post_id in text.extract_iter(
                    page, 'class="link-block" href="/posts/', '"'):
                yield post_id

            if not post_id:
                return
            params["page"] += 1
```

mikf commented on gallery_dl/extractor/fantia.py:
```diff
@@ -0,0 +1,149 @@
+        self.creator_id = match.group(1)
+
+    def posts(self):
+        base_url = self.root+"/fanclubs/"+self.creator_id+"/posts?page="
+
+        return self._pagination(base_url)
```
See previous comment
```suggestion
        url = "{}/fanclubs/{}/posts".format(self.root, self.creator_id)
        return self._pagination(url)
```

mikf commented on gallery_dl/extractor/fantia.py:
```diff
@@ -0,0 +1,149 @@
+        self.post_id = match.group(1)
+
+    def posts(self):
+        yield self._get_post_data(self.post_id)
```
See previous comment, and returning a 1-tuple is better than a yield/generator
```suggestion
        return (self.post_id,)
```

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,157 @@
+            final_post["num"] = num
+            num += 1
+            yield Message.Url, cover_image, final_post
+
```
Any code after this line only runs if `content_body` has anything in it, so we should do an early return here and simplify the checks below.
```suggestion

        if not content_body:
            return

```

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,157 @@
+            yield Message.Url, cover_image, final_post
+
+        for group in ("images", "imageMap"):
+            if group in (content_body or []):
```
```suggestion
            if group in content_body:
```

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,157 @@
+                    yield Message.Url, item["originalUrl"], final_post
+
+        for group in ("files", "fileMap"):
+            if group in (content_body or []):
```
```suggestion
            if group in content_body:
```

mikf commented on gallery_dl/extractor/fanbox.py:
```diff
@@ -0,0 +1,157 @@
+                "title"      : str,
+            },
+        }),
+    )
```
More tests to cover most URL possibilities
```suggestion
        ("https://xub.fanbox.cc/posts"),
        ("https://www.fanbox.cc/@xub/"),
        ("https://www.fanbox.cc/@xub/posts"),
    )
```

thatfuckingbird commented:
@abslamp Thanks, I will look at it in the coming days.
Can you put up a vimeo and soundcloud link too? Or actually just checking what the videoId for those looks like would be enough, so I can generate a URL from it

mikf reviewed (commented):
I've been playing around with this a bit more and found a few things in regards to URL patterns and fetching results. In summary:
- it should be possible safe a lot of requests to `https://api.fanbox.cc/post.info`
- use `self.request(url, params=params)` instead of building URLs yourself

mikf commented:
@thatfuckingbird  @abslamp There should also be something useful in the test data from PixivUtil2: https://github.com/Nandaka/PixivUtil2/tree/master/test

abslamp commented:
@thatfuckingbird SoundCloud and Vimeo links added. Seems that SoundCloud is not working correctly.

thatfuckingbird commented:
Added support for embedded videos and addressed the comments above.

Looking at the pixivutil test json files, I realized that article type posts and the imageMap/fileMap things are not handled correctly in the fanbox downloader. I also want to do another round of testing with the new changes. Going to do these soon.

mikf commented on docs/configuration.rst:
```diff
@@ -947,6 +947,23 @@ Description
     Download full-sized original images if available.
 
 
+extractor.fanbox.videos
+-----------------------
+Type
+    ``bool`` or ``string``
+Default
+    ``true``
+Description
+    Control behavior on videos embedded from external sites.
+    Recognizes embeds from YouTube, Vimeo and SoundCloud.
+
+    * ``true``: Extract video URLs (videos are not downloaded, as
+      galley-dl does not support these sites natively)
+    * ``"ytdl"``: Download videos and let `youtube-dl`_ handle all of
+      video extraction and download
+    * ``false``: Ignore videos
```
I think this option should be simplified to just
* `true`: Download embedded media with youtube-dl (what is currently happening for `"ytdl"`)
* `false`: Ignore videos

with just "Download embedded media from  YouTube, Vimeo, and SoundCloud with youtube-dl" as description.
The current `true` would download the HTML content of those external sites and I don't think anyone would want that.

thatfuckingbird replied:
Hmm, `true` shouldn't download html, instead the links show up in the unsupported URLs (the `--write-unsupported` flag),
though I can remove it, I don't particularly mind either way.

mikf replied:
You are right, they would land in the "unsupported" file. I didn't realize it's using`Message.Queue`.
But that wouldn't use youtube-dl to download them, either. It's `Message.Url` to download and `Message.Queue` to potentially spawn a new extractor.

I would still only have `true` or `false` as options (because its easier).
It'd need to use `Message.Url` for `ytdl:…` URIs if you want to keep all three.

thatfuckingbird commented:
OK, almost done. Looked at pixivutil and based on the examples found there I added support for the remaining post types:

-"entry" type posts just contain a "html" entry with raw html in it, have to extract images from that (I followed what pixivutil does here for exactly what type of image URLs are downloaded)

-"article" type posts seem to be similar in purpose but a newer format, it has paragraphs (+styling info, etc.) as json objects. This is where fileMap/imageMap is used. Pixivutil actually parses the paragraphs from the json and generates a HTML with the written post content, but I think it is better if gallery-dl just saves the whole article json into the metadata, then it can be postprocessed later by the user. Fortunately files/images can simply be downloaded based on the contents of imageMap/fileMap.

Handling embeds was also extended, turns out there are other types of embeds, some which gallery-dl can handle itself (e.g. twitter). I renamed the "videos" option to "embeds" but the idea is mostly the same. @abslamp You can delete the test posts now, thanks for the help.

Fortunately some of the posts that pixivutil tests use are public so I added some tests for the above.

Only one problem remains, that is the handling of Fanbox embeds (a fanbox post embedded in another). For some reason, if I yield a fanbox URL while parsing a fanbox post, it gets ignored (written into the unsupported urls file) despite the same code working with yielding twitter URLs (and if I manually run gallery-dl with the yielded Fanbox url then it is recognized correctly).
@mikf Can you take a look at this? The last test URL has a fanbox emvehttps://steelwire.fanbox.cc/posts/285502

mikf commented:
>  if I yield a fanbox URL while parsing a fanbox post, it gets ignored 

You need to specify the expected `Extractor` class as `_extractor` in the metadata dict:
``` python
             final_post["_extractor"] = FanboxCreatorExtractor
```

When there's no `_extractor` field or it is set to `None`, [`blacklist/whitelist`](https://github.com/mikf/gallery-dl/blob/master/docs/configuration.rst#extractorblacklist--whitelist) apply, which ignores URLs for the same `category` by default.

thatfuckingbird commented:
Thanks, that did it. I think everything's done now.

mo-han commented:
@thatfuckingbird 
Could you please add parsing and downloading of those "profile" and "plans" images?
Currently am saving those pictures by drag & drop, since they are just few, but it would be much more convenient if it's automated.
Thanks.
